File: internal/services/s3/sigv4.go
package s3

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"
)

const (
	sigV4Algorithm = "AWS4-HMAC-SHA256"
	sigV4Service   = "s3"
)

var nowUTC = func() time.Time {
	return time.Now().UTC()
}

type signatureError struct {
	code   string
	status int
}

func (e signatureError) Error() string {
	return e.code
}

func (s *Server) verifySignature(r *http.Request) error {
	hasPresign := r.URL.Query().Get("X-Amz-Algorithm") != ""
	hasHeaderAuth := r.Header.Get("Authorization") != ""
	if !hasPresign && !hasHeaderAuth {
		if strings.EqualFold(s.config.AuthMode, "strict") {
			return signatureError{code: "AccessDenied", status: http.StatusForbidden}
		}
		return nil
	}
	if hasPresign {
		return s.verifyPresignedURL(r)
	}
	return s.verifyAuthorizationHeader(r)
}

func (s *Server) verifyPresignedURL(r *http.Request) error {
	query := r.URL.Query()
	if query.Get("X-Amz-Algorithm") != sigV4Algorithm {
		return signatureError{code: "InvalidArgument", status: http.StatusBadRequest}
	}
	credential := query.Get("X-Amz-Credential")
	accessKey, dateStamp, region, service, ok := parseCredentialScope(credential)
	if !ok {
		return signatureError{code: "AuthorizationHeaderMalformed", status: http.StatusBadRequest}
	}
	if !s.validCredential(accessKey, region, service) {
		return signatureError{code: "InvalidAccessKeyId", status: http.StatusForbidden}
	}
	amzDate, err := time.Parse("20060102T150405Z", query.Get("X-Amz-Date"))
	if err != nil {
		return signatureError{code: "AccessDenied", status: http.StatusForbidden}
	}
	expires, err := strconv.ParseInt(query.Get("X-Amz-Expires"), 10, 64)
	if err != nil || expires < 0 || expires > 604800 {
		return signatureError{code: "AccessDenied", status: http.StatusForbidden}
	}
	if nowUTC().After(amzDate.Add(time.Duration(expires) * time.Second)) {
		return signatureError{code: "AccessDenied", status: http.StatusForbidden}
	}

	signedHeaders := query.Get("X-Amz-SignedHeaders")
	if signedHeaders == "" {
		return signatureError{code: "AuthorizationHeaderMalformed", status: http.StatusBadRequest}
	}
	expectedSignature := s.signatureForRequest(r, dateStamp, region, signedHeaders, "UNSIGNED-PAYLOAD", "X-Amz-Signature")
	if !hmac.Equal([]byte(query.Get("X-Amz-Signature")), []byte(expectedSignature)) {
		return signatureError{code: "SignatureDoesNotMatch", status: http.StatusForbidden}
	}
	return nil
}

func (s *Server) verifyAuthorizationHeader(r *http.Request) error {
	auth := r.Header.Get("Authorization")
	if !strings.HasPrefix(auth, sigV4Algorithm+" ") {
		return signatureError{code: "AuthorizationHeaderMalformed", status: http.StatusBadRequest}
	}
	values := parseAuthParams(strings.TrimPrefix(auth, sigV4Algorithm+" "))
	credential := values["Credential"]
	signedHeaders := values["SignedHeaders"]
	signature := values["Signature"]
	accessKey, dateStamp, region, service, ok := parseCredentialScope(credential)
	if !ok || signedHeaders == "" || signature == "" {
		return signatureError{code: "AuthorizationHeaderMalformed", status: http.StatusBadRequest}
	}
	if !s.validCredential(accessKey, region, service) {
		return signatureError{code: "InvalidAccessKeyId", status: http.StatusForbidden}
	}
	if r.Header.Get("x-amz-date") == "" {
		return signatureError{code: "AuthorizationHeaderMalformed", status: http.StatusBadRequest}
	}
	payloadHash := r.Header.Get("x-amz-content-sha256")
	if payloadHash == "" {
		payloadHash = "UNSIGNED-PAYLOAD"
	} else if err := verifyPayloadHash(r, payloadHash); err != nil {
		return err
	}
	expectedSignature := s.signatureForRequest(r, dateStamp, region, signedHeaders, payloadHash, "")
	if !hmac.Equal([]byte(signature), []byte(expectedSignature)) {
		return signatureError{code: "SignatureDoesNotMatch", status: http.StatusForbidden}
	}
	return nil
}

func (s *Server) validCredential(accessKey string, region string, service string) bool {
	configuredAccessKey := s.config.AccessKeyID
	if configuredAccessKey == "" {
		configuredAccessKey = "dev"
	}
	configuredRegion := s.config.Region
	if configuredRegion == "" {
		configuredRegion = "us-east-1"
	}
	return accessKey == configuredAccessKey && region == configuredRegion && service == sigV4Service
}

func verifyPayloadHash(r *http.Request, payloadHash string) error {
	if payloadHash == "UNSIGNED-PAYLOAD" {
		return nil
	}
	if strings.HasPrefix(payloadHash, "STREAMING-") {
		return signatureError{code: "NotImplemented", status: http.StatusNotImplemented}
	}
	body, err := io.ReadAll(r.Body)
	if err != nil {
		return signatureError{code: "InvalidRequest", status: http.StatusBadRequest}
	}
	_ = r.Body.Close()
	r.Body = io.NopCloser(bytes.NewReader(body))
	if got := sha256Hex(body); !hmac.Equal([]byte(strings.ToLower(payloadHash)), []byte(got)) {
		return signatureError{code: "XAmzContentSHA256Mismatch", status: http.StatusBadRequest}
	}
	return nil
}

func (s *Server) signatureForRequest(r *http.Request, dateStamp string, region string, signedHeaders string, payloadHash string, ignoredQueryKey string) string {
	canonicalRequest := strings.Join([]string{
		r.Method,
		canonicalURI(r.URL.Path),
		canonicalQueryString(r.URL.Query(), ignoredQueryKey),
		canonicalHeaders(r, signedHeaders),
		strings.ToLower(signedHeaders),
		payloadHash,
	}, "\n")
	amzDate := r.URL.Query().Get("X-Amz-Date")
	if amzDate == "" {
		amzDate = r.Header.Get("x-amz-date")
	}
	scope := strings.Join([]string{dateStamp, region, sigV4Service, "aws4_request"}, "/")
	stringToSign := strings.Join([]string{
		sigV4Algorithm,
		amzDate,
		scope,
		sha256Hex([]byte(canonicalRequest)),
	}, "\n")
	signature := hmacSHA256(deriveSigningKey(s.config.SecretAccessKey, dateStamp, region), stringToSign)
	return hex.EncodeToString(signature)
}

func parseCredentialScope(credential string) (accessKey string, dateStamp string, region string, service string, ok bool) {
	parts := strings.Split(credential, "/")
	if len(parts) != 5 || parts[4] != "aws4_request" {
		return "", "", "", "", false
	}
	return parts[0], parts[1], parts[2], parts[3], true
}

func parseAuthParams(value string) map[string]string {
	result := map[string]string{}
	for _, part := range strings.Split(value, ",") {
		key, val, ok := strings.Cut(strings.TrimSpace(part), "=")
		if ok {
			result[key] = val
		}
	}
	return result
}

func canonicalURI(path string) string {
	if path == "" {
		return "/"
	}
	return awsPercentEncode(path, "/~")
}

func canonicalQueryString(values url.Values, ignoredKey string) string {
	type pair struct {
		key   string
		value string
	}
	var pairs []pair
	for key, vals := range values {
		if key == ignoredKey {
			continue
		}
		if len(vals) == 0 {
			pairs = append(pairs, pair{key: key})
			continue
		}
		for _, val := range vals {
			pairs = append(pairs, pair{key: key, value: val})
		}
	}
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].key == pairs[j].key {
			return pairs[i].value < pairs[j].value
		}
		return pairs[i].key < pairs[j].key
	})
	encoded := make([]string, 0, len(pairs))
	for _, item := range pairs {
		encoded = append(encoded, awsPercentEncode(item.key, "~-_")+"="+awsPercentEncode(item.value, "~-_"))
	}
	return strings.Join(encoded, "&")
}

func canonicalHeaders(r *http.Request, signedHeaders string) string {
	var b strings.Builder
	for _, name := range strings.Split(strings.ToLower(signedHeaders), ";") {
		name = strings.TrimSpace(name)
		if name == "" {
			continue
		}
		value := r.Header.Get(name)
		if name == "host" {
			value = r.Host
		}
		fmt.Fprintf(&b, "%s:%s\n", name, normalizeHeaderValue(value))
	}
	return b.String()
}

func normalizeHeaderValue(value string) string {
	return strings.Join(strings.Fields(value), " ")
}

func awsPercentEncode(value string, safe string) string {
	var b strings.Builder
	for i := 0; i < len(value); i++ {
		c := value[i]
		if (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '~' || strings.ContainsRune(safe, rune(c)) {
			b.WriteByte(c)
			continue
		}
		fmt.Fprintf(&b, "%%%02X", c)
	}
	return b.String()
}

func deriveSigningKey(secret string, dateStamp string, region string) []byte {
	if secret == "" {
		secret = "dev"
	}
	dateKey := hmacSHA256([]byte("AWS4"+secret), dateStamp)
	regionKey := hmacSHA256(dateKey, region)
	serviceKey := hmacSHA256(regionKey, sigV4Service)
	return hmacSHA256(serviceKey, "aws4_request")
}

func hmacSHA256(key []byte, value string) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(value))
	return mac.Sum(nil)
}

func sha256Hex(value []byte) string {
	sum := sha256.Sum256(value)
	return hex.EncodeToString(sum[:])
}

func writeSignatureError(w http.ResponseWriter, err error) {
	if sigErr, ok := err.(signatureError); ok {
		writeXMLError(w, sigErr.code, sigErr.code, sigErr.status)
		return
	}
	writeXMLError(w, "AccessDenied", "AccessDenied", http.StatusForbidden)
}

File: services/applicationautoscaling/src/sigv4.rs
//! Mirrors `internal/services/applicationautoscaling/sigv4.rs`.
//!
//! AWS SigV4 verification with three modes (relaxed / signed-relaxed / strict),
//! matching the legacy server's reply names and HTTP statuses exactly.

use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256};

type HmacSha256 = Hmac<Sha256>;

pub const SIGV4_ALGORITHM: &str = "AWS4-HMAC-SHA256";
pub const SIGV4_SERVICE: &str = "application-autoscaling";

/// A signature verification failure, carrying the AWS error name and HTTP
/// status the legacy server would return.
#[derive(Debug, Clone)]
pub struct SignatureError {
    pub name: &'static str,
    pub status: u16,
}

fn err(name: &'static str, status: u16) -> SignatureError {
    SignatureError { name, status }
}

/// Minimal request view the verifier needs (headers + method + path + query +
/// host + body), so it stays independent of the HTTP framework.
pub struct SignedRequest<'a> {
    pub method: &'a str,
    pub path: &'a str,
    pub query: &'a str,
    pub host: &'a str,
    pub authorization: &'a str,
    pub amz_date: &'a str,
    pub content_sha256: &'a str,
    /// Resolves a header value by lowercase name (for canonical headers).
    pub header: &'a dyn Fn(&str) -> Option<String>,
    pub body: &'a [u8],
}

pub struct Credentials<'a> {
    pub auth_mode: &'a str,
    pub access_key_id: &'a str,
    pub secret_access_key: &'a str,
    pub region: &'a str,
}

fn default_str<'a>(value: &'a str, fallback: &'a str) -> &'a str {
    if value.is_empty() {
        fallback
    } else {
        value
    }
}

/// Mirrors `Server.verifySignature`.
pub fn verify_signature(req: &SignedRequest, creds: &Credentials) -> Result<(), SignatureError> {
    if creds.auth_mode.is_empty() || creds.auth_mode.eq_ignore_ascii_case("relaxed") {
        return Ok(());
    }
    let auth = req.authorization.trim();
    if auth.is_empty() {
        return Err(err("AccessDeniedException", 403));
    }
    if creds.auth_mode.eq_ignore_ascii_case("signed-relaxed") {
        return verify_shape(req, auth);
    }
    verify_full(req, auth, creds)
}

fn verify_full(req: &SignedRequest, auth: &str, creds: &Credentials) -> Result<(), SignatureError> {
    let prefix = format!("{SIGV4_ALGORITHM} ");
    if !auth.starts_with(&prefix) {
        return Err(err("IncompleteSignatureException", 400));
    }
    let values = parse_auth_params(&auth[prefix.len()..]);
    let credential = values.get("Credential").cloned().unwrap_or_default();
    let signed_headers = values.get("SignedHeaders").cloned().unwrap_or_default();
    let signature = values.get("Signature").cloned().unwrap_or_default();
    let (access_key, date_stamp, region, service) = match parse_credential_scope(&credential) {
        Some(v) => v,
        None => return Err(err("IncompleteSignatureException", 400)),
    };
    if signed_headers.is_empty() || signature.is_empty() {
        return Err(err("IncompleteSignatureException", 400));
    }
    if !valid_credential(&access_key, &region, &service, creds) {
        return Err(err("UnrecognizedClientException", 403));
    }
    if req.amz_date.is_empty() {
        return Err(err("IncompleteSignatureException", 400));
    }
    let payload_hash = if req.content_sha256.is_empty() {
        "UNSIGNED-PAYLOAD".to_string()
    } else {
        verify_payload_hash(req, req.content_sha256)?;
        req.content_sha256.to_string()
    };
    let expected = signature_for_request(
        req,
        &date_stamp,
        &region,
        &signed_headers,
        &payload_hash,
        creds,
    );
    if !constant_time_eq(signature.as_bytes(), expected.as_bytes()) {
        return Err(err("InvalidSignatureException", 403));
    }
    Ok(())
}

fn verify_shape(req: &SignedRequest, auth: &str) -> Result<(), SignatureError> {
    let prefix = format!("{SIGV4_ALGORITHM} ");
    if !auth.starts_with(&prefix) {
        return Err(err("IncompleteSignatureException", 400));
    }
    let values = parse_auth_params(&auth[prefix.len()..]);
    let credential = values.get("Credential").cloned().unwrap_or_default();
    let signed_headers = values.get("SignedHeaders").cloned().unwrap_or_default();
    let signature = values.get("Signature").cloned().unwrap_or_default();
    let (_, _, _, service) = match parse_credential_scope(&credential) {
        Some(v) => v,
        None => return Err(err("IncompleteSignatureException", 400)),
    };
    if signed_headers.is_empty() || signature.is_empty() {
        return Err(err("IncompleteSignatureException", 400));
    }
    if service != SIGV4_SERVICE || !is_lower_hex(&signature, 64) {
        return Err(err("IncompleteSignatureException", 400));
    }
    if req.amz_date.is_empty() {
        return Err(err("IncompleteSignatureException", 400));
    }
    if !req.content_sha256.is_empty() {
        return verify_payload_hash(req, req.content_sha256);
    }
    Ok(())
}

fn is_lower_hex(value: &str, length: usize) -> bool {
    value.len() == length
        && value
            .bytes()
            .all(|c| c.is_ascii_digit() || (b'a'..=b'f').contains(&c))
}

fn valid_credential(access_key: &str, region: &str, service: &str, creds: &Credentials) -> bool {
    let configured_access_key = default_str(creds.access_key_id, "dev");
    let configured_region = default_str(creds.region, "us-east-1");
    access_key == configured_access_key && region == configured_region && service == SIGV4_SERVICE
}

fn verify_payload_hash(req: &SignedRequest, payload_hash: &str) -> Result<(), SignatureError> {
    if payload_hash == "UNSIGNED-PAYLOAD" {
        return Ok(());
    }
    if payload_hash.starts_with("STREAMING-") {
        return Err(err("NotImplemented", 501));
    }
    let got = sha256_hex(req.body);
    if !constant_time_eq(payload_hash.to_ascii_lowercase().as_bytes(), got.as_bytes()) {
        return Err(err("InvalidSignatureException", 403));
    }
    Ok(())
}

fn signature_for_request(
    req: &SignedRequest,
    date_stamp: &str,
    region: &str,
    signed_headers: &str,
    payload_hash: &str,
    creds: &Credentials,
) -> String {
    let canonical_request = [
        req.method,
        &canonical_uri(req.path),
        &canonical_query_string(req.query),
        &canonical_headers(req, signed_headers),
        &signed_headers.to_ascii_lowercase(),
        payload_hash,
    ]
    .join("\n");
    let scope = [date_stamp, region, SIGV4_SERVICE, "aws4_request"].join("/");
    let string_to_sign = [
        SIGV4_ALGORITHM,
        req.amz_date,
        &scope,
        &sha256_hex(canonical_request.as_bytes()),
    ]
    .join("\n");
    let signing_key = derive_signing_key(creds.secret_access_key, date_stamp, region);
    hex::encode(hmac_sha256(&signing_key, string_to_sign.as_bytes()))
}

fn parse_credential_scope(credential: &str) -> Option<(String, String, String, String)> {
    let parts: Vec<&str> = credential.split('/').collect();
    if parts.len() != 5 || parts[4] != "aws4_request" {
        return None;
    }
    Some((
        parts[0].to_string(),
        parts[1].to_string(),
        parts[2].to_string(),
        parts[3].to_string(),
    ))
}

fn parse_auth_params(value: &str) -> std::collections::HashMap<String, String> {
    let mut result = std::collections::HashMap::new();
    for part in value.split(',') {
        if let Some((k, v)) = part.trim().split_once('=') {
            result.insert(k.to_string(), v.to_string());
        }
    }
    result
}

fn canonical_uri(path: &str) -> String {
    if path.is_empty() {
        return "/".to_string();
    }
    aws_percent_encode(path, "/~")
}

fn canonical_query_string(query: &str) -> String {
    if query.is_empty() {
        return String::new();
    }
    let mut pairs: Vec<(String, String)> = Vec::new();
    for item in query.split('&') {
        match item.split_once('=') {
            Some((k, v)) => pairs.push((url_decode(k), url_decode(v))),
            None => pairs.push((url_decode(item), String::new())),
        }
    }
    pairs.sort_by(|a, b| a.0.cmp(&b.0).then(a.1.cmp(&b.1)));
    pairs
        .iter()
        .map(|(k, v)| {
            format!(
                "{}={}",
                aws_percent_encode(k, "~-_"),
                aws_percent_encode(v, "~-_")
            )
        })
        .collect::<Vec<_>>()
        .join("&")
}

fn canonical_headers(req: &SignedRequest, signed_headers: &str) -> String {
    let mut out = String::new();
    for name in signed_headers.to_ascii_lowercase().split(';') {
        let name = name.trim();
        if name.is_empty() {
            continue;
        }
        let value = if name == "host" {
            req.host.to_string()
        } else {
            (req.header)(name).unwrap_or_default()
        };
        out.push_str(name);
        out.push(':');
        out.push_str(&normalize_header_value(&value));
        out.push('\n');
    }
    out
}

fn normalize_header_value(value: &str) -> String {
    value.split_whitespace().collect::<Vec<_>>().join(" ")
}

fn aws_percent_encode(value: &str, safe: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for &c in value.as_bytes() {
        let ch = c as char;
        if ch.is_ascii_alphanumeric()
            || ch == '-'
            || ch == '_'
            || ch == '.'
            || ch == '~'
            || safe.contains(ch)
        {
            out.push(ch);
        } else {
            out.push_str(&format!("%{:02X}", c));
        }
    }
    out
}

fn url_decode(s: &str) -> String {
    let bytes = s.as_bytes();
    let mut out = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        match bytes[i] {
            b'%' if i + 2 < bytes.len() => {
                if let Ok(b) = u8::from_str_radix(&s[i + 1..i + 3], 16) {
                    out.push(b);
                    i += 3;
                    continue;
                }
                out.push(bytes[i]);
                i += 1;
            }
            b'+' => {
                out.push(b' ');
                i += 1;
            }
            b => {
                out.push(b);
                i += 1;
            }
        }
    }
    String::from_utf8_lossy(&out).into_owned()
}

fn derive_signing_key(secret: &str, date_stamp: &str, region: &str) -> Vec<u8> {
    let secret = default_str(secret, "dev");
    let date_key = hmac_sha256(format!("AWS4{secret}").as_bytes(), date_stamp.as_bytes());
    let region_key = hmac_sha256(&date_key, region.as_bytes());
    let service_key = hmac_sha256(&region_key, SIGV4_SERVICE.as_bytes());
    hmac_sha256(&service_key, b"aws4_request")
}

fn hmac_sha256(key: &[u8], value: &[u8]) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC accepts any key length");
    mac.update(value);
    mac.finalize().into_bytes().to_vec()
}

fn sha256_hex(value: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(value);
    hex::encode(hasher.finalize())
}

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}
